metaflow/plugins/kfp/kfp_split_contexts.py

```diff
@@ -0,0 +1,178 @@
+                self.graph[parent_context_step_name].type == "foreach"
+                and parent_context_step_name
+                == current_node.in_funcs[0]  # and is a direct parent
+            ):
+                # if the direct parent node is a foreach
+                # and current_node.is_inside_foreach (we are in a foreach) then:
+                #   it's context_node_task_id is passed_in_split_indexes
+                #   minus the last split_index which is for the inner loop
+                split_indices_but_last_one = passed_in_split_indexes.split(
+                    SPLIT_INDEX_SEPARATOR
+                )[:-1]
+                context_split_indexes = SPLIT_INDEX_SEPARATOR.join(
+                    split_indices_but_last_one
+                )
+            else:
+                context_split_indexes = passed_in_split_indexes
+
+            context_node_task_id = f"{context_node_task_id}.{context_split_indexes}"
+        else:
+            # not is_inside_foreach, hence context_node_task_id is None
+            # and the step_kfp_output_path doesn't have task_id in it.
+            pass
+
+        self.logger(f"context_node_task_id: {context_node_task_id}")
+        step_kfp_output_path = KfpSplitContext._build_step_kfp_output_path(
+            self.flow_root, parent_context_step_name, context_node_task_id
+        )
+
+        with S3() as s3:
+            self.logger(
+                f"get_parent_context({parent_context_step_name}: {step_kfp_output_path}"
+            )
+            input_context = json.loads(s3.get(step_kfp_output_path).text)
+            self.logger(pprint.pformat(input_context), head="--")
+            return input_context
+
+    def get_current_step_split_index(self, passed_in_split_indexes: str) -> str:
+        if self.node.is_inside_foreach:
+            # the index is the last appended split ordinal
+            return passed_in_split_indexes.split(SPLIT_INDEX_SEPARATOR)[-1]
+        else:
+            return ""
+
+    @staticmethod
+    def save_context_to_local_fs(context_dict: Dict[str, str]):
+        """
+        Used by kfp_decorator.py to save the context to disk.
+        step_op_func opens this file to read out and return foreach_splits
+        """
+        # write: context_dict to local FS to return
+        with open(KFP_METAFLOW_CONTEXT_DICT_PATH, "w") as file:
+            json.dump(context_dict, file)
+
+    def upload_context_to_flow_root(self, context_dict: Dict[str, str]):
+        # upload: context_dict
+        with S3() as s3:
+            step_kfp_output_path = KfpSplitContext._build_step_kfp_output_path(
+                self.flow_root, self.step_name, current.task_id
+            )
+            s3.put(step_kfp_output_path, json.dumps(context_dict))
+
+    @staticmethod
+    def get_step_task_id(task_id: str, passed_in_split_indexes: str) -> str:
+        return f"{task_id}.{passed_in_split_indexes}".strip(".")
+
+    @staticmethod
+    def _build_step_kfp_output_path(
```

Comment:
Seems strange to me to have a private static method, mind just making this public or a non-staticmethod?